[W5][T09-1] Lee Yu Choy

lycjackie opened the pull request:
Add a new `sort` command

Previously we could only list all person in the address book, but there was no order.

Let's create a new `sort` command to sort all person in the address book by their name.

Commit:
Add SortCommand class for sort feature

Commit:
Add comparator for Name class

Currently there is no comparator for the `Name` class to do any
comparision

Let's implement a comparator for the class so that we can use it
to sort in the future

Commit:
Add sort function for UniquePersonList

Sort the list of person based on their names in alphabetical order

Commit:
Add sort feature for AddressBook

Previously we have implemented the sort feature for UniquePersonList,
however there was no implemented of it in the actual AddressBook

Let's now call implement the sort feature into AddressBook class
so the list of allPerson can now be sorted

Commit:
Add implemented for SortCommand

Commit:
Add 'sort' keyword to parser

AddressBook now can execute `sort`

Commit:
Update help command to reflect changes

Commit:
Update I/O test cases

Reorder some input in adding person so that we can see the sorting

Update the expected result to reflect accordingly

Commit:
Add SortCommandTest with 4 JUnits tests

Slight modification made to TypicalPersons class to fit tests

Commit:
Update UserGuide to reflect sort command

Creastery commented on src/seedu/addressbook/commands/SortCommand.java:
```diff
@@ -0,0 +1,25 @@
+
+
+/**
+ * Lists all persons in the address book to the user.
```
Class comment does not match description of the command!

chyeo replied:
The description is the same as that of the ListCommand.java class

lycjackie replied:
> The description is the same as that of the ListCommand.java class

Fixed in latest commit.

Creastery commented on src/seedu/addressbook/data/person/Name.java:
```diff
@@ -60,4 +60,9 @@ public int hashCode() {
+    @Override
+    public int compareTo(Name name){
+        return this.fullName.compareToIgnoreCase(name.fullName);
+    }
```
Perhaps it would be good to amend commit message to indicate that the default comparator imposes lexicographical order and is case-insensitive!

Creastery replied:
Line 64 needs a whitespace before `{`:

> Java reserved words should be followed by a white space.

Refer to [CS2113T Java Coding Standard](https://nuscs2113-ay1819s2.github.io/website/coding-standards/java/intermediate.html) - White space within a statement section.

Creastery commented on src/seedu/addressbook/data/person/UniquePersonList.java:
```diff
@@ -81,6 +82,19 @@ public UniquePersonList(UniquePersonList source) {
+            Name personBName = personB.getName();
+            return personAName.compareTo(personBName);
+        };
+        Collections.sort(internalList,personComparator);
```
Line 96 needs a whitespace after `,`:

> Commas should be followed by a white space.

Refer to [CS2113T Java Coding Standard](https://nuscs2113-ay1819s2.github.io/website/coding-standards/java/intermediate.html) - White space within a statement section.

Creastery commented on src/seedu/addressbook/data/AddressBook.java:
```diff
@@ -54,6 +54,13 @@ public void removePerson(ReadOnlyPerson toRemove) throws PersonNotFoundException
+    /**
+     * Sorts all person in the address book by their name in alphabetical order.
+     */
+    public void sortPerson(){
```
Line 60 needs a whitespace before `{`:

> Java reserved words should be followed by a white space.

Refer to [CS2113T Java Coding Standard](https://nuscs2113-ay1819s2.github.io/website/coding-standards/java/intermediate.html) - White space within a statement section.

Creastery commented on src/seedu/addressbook/parser/Parser.java:
```diff
@@ -1,5 +1,6 @@
 package seedu.addressbook.parser;
 
+import seedu.addressbook.commands.SortCommand;
```
This import statement should be moved down to after Line 23.
> The ordering of import statements must be consistent.

Refer to [CS2113T Java Coding Standard](https://nuscs2113-ay1819s2.github.io/website/coding-standards/java/intermediate.html) - The ordering of import statements must be consistent section.

Creastery commented on src/seedu/addressbook/commands/SortCommand.java:
```diff
@@ -0,0 +1,25 @@
+
+    public static final String MESSAGE_USAGE = COMMAND_WORD
+            + ": Displays all persons in the address book as a list sorted by their name"
+            + ", in alphabetical order.\n"
```
According to your implementation:
https://github.com/nusCS2113-AY1819S2/addressbook-level2/blob/7942d0695d9c32e501ba9fda7df80a080db31127/src/seedu/addressbook/data/person/Name.java#L63-L66

You are sorting the list in **lexicographical order**, not **alphabetical order**.

Refer to [this relevant Quora question](https://www.quora.com/Whats-the-difference-between-lexicographical-and-alphabetical-order) for more information.

Creastery commented on test/expected.txt:
```diff
@@ -19,6 +19,8 @@
 || Example: find alice bob charlie
 || list: Displays all persons in the address book as a list with index numbers.
 || Example: list
+|| sort: Displays all persons in the address book as a list sorted by their name, in alphabetical order.
```
Same as https://github.com/nusCS2113-AY1819S2/addressbook-level2/pull/17#discussion_r255319744.

Creastery commented on test/java/seedu/addressbook/commands/SortCommandTest.java:
```diff
@@ -0,0 +1,60 @@
+
+import org.junit.Test;
+
+import static org.junit.Assert.assertEquals;
```
Same as https://github.com/nusCS2113-AY1819S2/addressbook-level2/pull/17#discussion_r255319352.

Creastery commented on docs/UserGuide.adoc:
```diff
@@ -72,6 +72,11 @@ Examples:
 
+== Listing all persons by their name : `sort`
+
+Shows a list of all persons sorted by their name in alphabetical orders, in the address book. +
```
Same as https://github.com/nusCS2113-AY1819S2/addressbook-level2/pull/17#discussion_r255319744.

Creastery reviewed (commented):
> Previously we have implemented the sort feature for UniquePersonList,
> however there was no implemented of it in the actual AddressBook
>
> Let's now call implement the sort feature into AddressBook class
> so the list of allPerson can now be sorted

I was left slightly confused by your commit message after reading this line in particular:

> ... so the list of allPerson can now be sorted

`sortPerson()` was never called in the program, so claiming that the list of `allPerson` can _now_ be sorted is rather misleading...

Creastery commented:
Do consider renaming the commits summaries to make the objective of the commits even clearer.

For instance,
- [7942d06]
Original: Add comparator for Name class
Suggestion: `Name` class: add lexicographic-order comparator
- [ea28f61] 
Original: Update help command to reflect changes
Suggestion: `HelpCommand` class: add `sort` command usage
- [9bb0223]
Original: Add SortCommandTest with 4 JUnits tests
Suggestion: JUnit: add `SortCommandTest` class to test `sort` command

Creastery commented:
Overall, really great work done to make small, incremental changes to complete one feature! 👍

Commit:
Fix check style

@Creastery have highlighted a lot of my checkstyle mismatch

Lets fix all of them

lycjackie commented:
> Overall, really great work done to make small, incremental changes to complete one feature! 👍

Thanks, @Creastery for the feedback! 
I have updated all the check style and comments issues!

Creastery commented on src/seedu/addressbook/data/person/Name.java:
```diff
@@ -60,8 +60,11 @@ public int hashCode() {
     @Override
-    public int compareTo(Name name){
+    public int compareTo(Name name) {
         return this.fullName.compareToIgnoreCase(name.fullName);
```
Apologies for missing this out earlier: 
If you are doing case-insensitive comparison of strings, perhaps it would be better to mention that the names are sorted using **case-insensitive** lexicographical order!

sanjukta99 commented on test/expected.txt:
```diff
@@ -93,33 +95,33 @@
-|| New person added: Dickson Ee Phone: 444444 Email: (private) dickson@nus.edu.sg Address: 444, delta street Tags: [friends]
+|| Enter command: || [Command entered:  add Betsy Choo pp/222222 pe/benchoo@nus.edu.sg pa/222, beta street t/secretive]
+|| New person added: Betsy Choo Phone: (private) 222222 Email: (private) benchoo@nus.edu.sg Address: (private) 222, beta street Tags: [secretive]
 || ===================================================
```
Good job on updating test cases!

sanjukta99 commented on docs/UserGuide.adoc:
```diff
@@ -72,6 +72,11 @@ Examples:
+Shows a list of all persons sorted by their name in lexicographical orders, in the address book. +
+Format: `sort`
+
 == Finding all persons containing any keyword in their name: `find`
```
Good job on updating the User Guide

sanjukta99 commented on src/seedu/addressbook/commands/SortCommand.java:
```diff
@@ -0,0 +1,27 @@
+/**
+ * Sort all persons in the address book in lexicographical order.
+ */
+public class SortCommand extends Command {
```
Perhaps you can mention that this command sorts in a *case-insensitive* lexicographical order?

sanjukta99 commented on test/java/seedu/addressbook/commands/SortCommandTest.java:
```diff
@@ -0,0 +1,61 @@
+     * Executes the sort command for the given list and verifies it is sorted
+     * the result matches the persons in the expectedPersons exactly in the same orders.
+     */
+
```
Great job on attempting to do some JUnit Testing!

sanjukta99 reviewed (commented):
Overall, a very good enhancement with some testing and relevant updates to the UG! Please read the reviews and close this PR.
